### PR TITLE
[cas] Implement file reading

### PR DESCRIPTION
**Pull request by nodirg**
This commit implements concurrent FS walk and reading of regular files
(except large files). After reading, the files are scheduled for a
presence check, which is not implemented in this CL.
Symlink support TBD.

### CODE REVIEW

File: go/pkg/cas/client.go
Comment (atetubou) [inline go/pkg/cas/client.go:40]: better to follow
https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences
?

File: go/pkg/cas/client.go
Comment (atetubou) [inline go/pkg/cas/client.go:71]: why choose 64 then?

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:49]:
```suggestion
type FilePredicate func(absPath string, mode os.FileMode) bool
```

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:89]: Is it fine to use eg here?

```suggestion
				eg.Go(func() error {
```

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:114]: https://golang.org/pkg/sync/#Map says
```
The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.
```

Can we start from map with mutext?

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:133]: better to wrap error if it come from standard library?

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:139]: ditto

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:153]:
```suggestion
	AbsPath      string
```

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:204]:
```suggestion
func (u *uploader) visitRegularFile(ctx context.Context, absPath string, info os.FileInfo) (*repb.FileNode, error) {
```

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:273]: nit
```suggestion
	// item.Open will reopen the file.
```

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:294]: better to wrap error from standard library?

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:300]: ditto

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:41]: Why not have this in uploader instead?

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:160]: Can we use node for this?
```suggestion
	node proto.Message // *repb.FileNode, *repb.DirectoryNode or *repb.SymlinkNode.
```

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:422]: Better to show digest instead of hash?

File: go/pkg/cas/upload_test.go
Comment (atetubou) [inline go/pkg/cas/upload_test.go:21]:
```suggestion
	absTestData, err := filepath.Abs("testdata")
```

File: go/pkg/cas/upload_test.go
Comment (atetubou) [inline go/pkg/cas/upload_test.go:155]:
```suggestion
	inputC := make(chan *UploadInput, len(inputs))
```

File: go/pkg/cas/upload_test.go
Comment (atetubou) [inline go/pkg/cas/upload_test.go:101]: no need to have this explicitly?

**Review by atetubou — CHANGES_REQUESTED** [review]

**Comment by atetubou** [comment]
cc @Jwata

File: go/pkg/cas/client.go
Comment (nodirg) [reply go/pkg/cas/client.go:40]: Done

File: go/pkg/cas/client.go
Comment (nodirg) [reply go/pkg/cas/client.go:71]: Changed to 32 and a TODO to tune it.

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:89]: done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:49]: done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:114]: switched to mutex and add a TODO to consider sync.Mutex

Comment (ramymedhat) [reply go/pkg/cas/upload.go:114]: "The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow ...". I think syncMap is appropriate for this use case, but in general this seems very close to singleflightcache. Would it make sense to use the same package for the same functionality instead of having different implementations? We can address performance concerns in the singleflightcache itself and just use it here.

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:133]: Done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:139]: Done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:153]: Done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:204]: Done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:273]: Done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:160]: Done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:422]: done

File: go/pkg/cas/upload_test.go
Comment (nodirg) [reply go/pkg/cas/upload_test.go:21]: Done

File: go/pkg/cas/upload_test.go
Comment (nodirg) [reply go/pkg/cas/upload_test.go:155]: Done

File: go/pkg/cas/upload_test.go
Comment (nodirg) [reply go/pkg/cas/upload_test.go:101]: sure

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:41]: > Does the use case of isolate have many cases of duplicate traversals of directories and use of input exclusions?

Yes. We often traverse the same directories, and there is almost always a generic regexp-based filter. 
I am no sure how to avoid this.

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:245]: done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:114]: @ramymedhat just to see if we are thinking about the same thing: do you mean that reusing a common functionality could save us 3 lines of code?

```go
	entryRaw, _ := u.fsCache.LoadOrStore(cacheKey, &fsCacheValue{})
	e := entryRaw.(*fsCacheValue)
	e.compute.Do(func() {
```

Is this over-generalization?

Also, FWIU the only usage of `singleflightcache` is for CAS, so if we switch to this package, do we still need `singleflightpackage`?

File: go/pkg/cas/upload.go
Comment (ramymedhat) [reply go/pkg/cas/upload.go:41]: Ah I see. Good to know.

I'm also concerned about the practicality of users using the same predicate function. Specially since the `client` package operates on the InputExclusions struct. So naturally it will be creating new functions when calling `cas` each time an execution comes in. Or it would maintain a cache of `InputExclusion` to predicate function.

Do you think it makes sense to defer the predicate cache until we start measuring performance? We may also get more insight once `client` starts using `cas`.  Alternatively we could put more restrictions on InputExclusions to  enforce there is at most one per input type, and then we can cache the actual regexs.

File: go/pkg/cas/upload.go
Comment (ramymedhat) [reply go/pkg/cas/upload.go:114]: My point is more about reusing the common lib to make sure whatever optimizations we do apply here and elsewhere. For instance, I think here you're doing one less lock which would be beneficial in the general implementation. Also with a central cache we can track its size and memory consumption, get cache hit rates and other stats.

Comment (ramymedhat) [reply go/pkg/cas/upload.go:114]: > Also, FWIU the only usage of singleflightcache is for CAS, so if we switch to this package, do we still need singleflightpackage?

We use the cache for other use cases besides file digest. We also use the file metadata cache outside of the SDK.

Comment (rubensf) [reply go/pkg/cas/upload.go]: +1 to what Ramy said. This can be useful outside of the SDK (to lookup metadata/digest checking). Enhancing `SingleFlightCache` (which likely can perform better) or eventually (in a later PR?) exporting this both sound good to me.

Comment (nodirg) [reply go/pkg/cas/upload.go]: > We also use the file metadata cache outside of the SDK.

Thanks, this didn't occur to me.

I am happy to improve singleflightcache in a separate CL (otherwise seems to be a scope creep). I just tried, but the semantics is not the same: `singleflightcache` avoids caching the error (if `valFn` errors out, then the value isn't stored), so the same trick doesn't work. `uploader` here doesn't need that, and thus can be simpler (just one map, with `once`s) and more efficient. Ideas?

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go]: Yes, it doesn't work on closures. I've removed predicate from this CL. We can revisit this in another CL.

For context, the performance [was already measured](https://bugs.chromium.org/p/chromium/issues/detail?id=1193375#c37).

`client` package isn't the only user of this package. In our package where we'd use `cas` package, we have a single regex per `UploadItem`, so it is easy to cache functions per regex. [Here is our code](https://chromium-review.googlesource.com/c/infra/luci/luci-go/+/2837288/1/client/cmd/isolate/lib/common.go#349).

Ramy, also could you clarify what exactly is the suggestion? If we don't have predicate in the cache key, then the code will become incorrect because one `UploadItem` may have one predicate, and another might have another. So their dir nodes shouldn't be the same. If you are suggesting to move the predicate from `UploadItem` to a parameter of `Upload()`, then it means separate uploads cannot reuse each other's cache.

> Do we have use cases besides file name regexes + file mode checks?
Chrome doesn't have use cases besides file name regexes.

File: go/pkg/cas/client.go
Comment (nodirg) [reply go/pkg/cas/client.go]: Updated this code. PTAL

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go]: added, although I don't see how it would help. There is only one place where we lock/unlock.

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:253]: added a reference to the doc.

File: go/pkg/cas/ioutil.go
Comment (nodirg) [reply go/pkg/cas/ioutil.go]: Done.

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:330]: added.

File: go/pkg/cas/client.go
Comment (rubensf) [inline go/pkg/cas/client.go:52]: nit: no newline

File: go/pkg/cas/upload.go
Comment (rubensf) [inline go/pkg/cas/upload.go:284]: wdyt of moving this function into its own to reduce indentation? no strong feelings.

File: go/pkg/cas/upload.go
Comment (rubensf) [inline go/pkg/cas/upload.go:308]: thinking: filepath functions _always_ call `Clean`, even when we know it isn't necessary (absPath should have been cleaned on `startProcessing`; AFAICT `info.Name()` should be as good as clean?). 

At the same time using `strings.Join(..., os.Separator)` feels awkward? But that's what filepath does: https://golang.org/src/path/filepath/path_unix.go. wdyt of using it directly?

File: go/pkg/cas/upload.go
Comment (rubensf) [reply go/pkg/cas/upload.go:330]: I meant on the error message :)

File: go/pkg/cas/upload.go
Comment (rubensf) [reply go/pkg/cas/upload.go]: I *think* those semantics aren't necessary. We might just need the stat metadata saved, which can be done at a later PR. @ramymedhat ?

**Review by rubensf — APPROVED** [review]
LGTM overall. Ramy and Takuto might have more thoughts.

File: go/pkg/cas/client.go
Comment (nodirg) [reply go/pkg/cas/client.go:52]: done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go]: Added a new func. PTAL. `strings.Join` seems overkill here.

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:284]: I considered that, but it means I'd also have to carry all the state (mutex, dir, wg, etc) via params, or maybe a separate struct, so overall it gets more complicated.

File: go/pkg/cas/upload.go
Comment (atetubou) [inline go/pkg/cas/upload.go:42]: This is not used in this PR now.

Comment (nodirg) [reply go/pkg/cas/upload.go]: removed all predicate-related stuff from this CL

**Comment by nodirg** [comment]
@rubensf idk how to respond to your resolved comment, but I've addressed it.

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:283]: done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:289]: done

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go]: WDYM by this?

Comment (nodirg) [reply go/pkg/cas/upload.go]: Given that predicate support is removed from this CL, I assume this discussion is resolved. Unresolve otherwise.

**Review by atetubou — APPROVED** [review]

File: go/pkg/cas/upload.go
Comment (ramymedhat) [reply go/pkg/cas/upload.go]: Yeah I think we don't need that. Ok to defer to a later PR though.

File: go/pkg/cas/upload.go
Comment (ramymedhat) [reply go/pkg/cas/upload.go]: Just responding to the last message:

> For context, the performance was already measured.
I meant overhead of applying the regex repeatedly on duplicate directories in practice. I couldn't find a specific test for that in the bug but I might've missed it.

My suggestions were to (1) make the cache contain an unfiltered list of items (files, subdirs, symlinks) and apply the predicate before uploading not before storing in cache. So the cache becomes effectively a ReadDir cache. Option (2) is to cache the regexes assuming we enforce there is a constant number of regexes in the InputExclusions slice. I agree we can't move the predicate to `Upload()`. But generally agree this makes sense in a separate PR.

**Review by ramymedhat — APPROVED** [review]

File: go/pkg/cas/upload.go
Comment (rubensf) [reply go/pkg/cas/upload.go:281]: I have a slight preference towards the many-arguments uglyness instead of 6 nesting levels uglyness. IMO it's easier to human parse the first than the second. But tbh this is pretty subjective, so I'm fine with whatever you think is best.

**Review by rubensf — APPROVED** [review]
FYI, if you want to reply to outdated comments, you'll have to go on commits -> click on hash of the commit where the comment was create -> add comment there.

Sadly, whenever you delete your branch after the PR is merged, checking those historical comments become pretty hard, since we squash the PRs and the intermediate comments aren't stored on github.

... github's interface is quite inferior to gerrit :(

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go:281]: I agree that 6 sounds a bit too much, but in this case moving it out to a separate top-level function would save us only one level of indentation :)

File: go/pkg/cas/upload.go
Comment (nodirg) [reply go/pkg/cas/upload.go]: I sent out https://github.com/bazelbuild/remote-apis-sdks/pull/303, but I couldn't figure out how to rebase this PR on top of the new one (something that I do in gerrit all the time). So I am going to have to merge this PR, and then update https://github.com/bazelbuild/remote-apis-sdks/pull/303 to use `singleflightcache` in `cas` package.

**Pull request merged**
